### PR TITLE
Updating TextHint accessibility

### PR DESCRIPTION
**Pull request by ychoi1**
### Context (Links to Quips/Jira)
https://instacart.atlassian.net/browse/CS-7397
https://instacart.atlassian.net/browse/CS-7398

### Problem
Text hints for `MaskedTextField` are not accessibility compliant:
- Text hint was able to receive focus when not visible on screen
- Text hint was not associated to relative field
- Validation error was not associated to relative field

### Solution
- Added `visible: hidden / visible` css to TextFieldHint
- Added `aria-describedby` to `MaskedTextField`. Passing hint and validation error input id

### CODE REVIEW

File: src/components/Forms/MaskedTextField.js [hunk 1]
```diff
@@ -285,6 +290,7 @@ class MaskedTextField extends React.Component {
             name={name}
             aria-required={required}
             aria-invalid={hasError}
+            aria-describedby={hasError ? `hint_${inputId} error_${inputId}` : `hint_${inputId}`}
```
Comment (wfro) [inline src/components/Forms/MaskedTextField.js]: Does a screenreader read the text from both of these elements if there's an error?

File: src/components/Forms/TextFieldHint.js [hunk 2]
```diff
@@ -15,9 +15,11 @@ const styles = {
     transition: 'opacity 250ms cubic-bezier(0.23, 1, 0.32, 1) 0ms',
     pointerEvents: 'none',
     zIndex: 1,
+    visibility: 'hidden',
```
Comment (wfro) [inline src/components/Forms/TextFieldHint.js]: Just to double check here, is this specifically so that screen readers won't read this content?

File: src/components/Forms/MaskedTextField.js [hunk 3]
```diff
@@ -275,7 +275,12 @@ class MaskedTextField extends React.Component {
             snacksTheme={snacksTheme}
           />
```
Comment (wfro) [inline src/components/Forms/MaskedTextField.js]: Is `MaskedTextField` the only text field that needs this? There's another `TextField` component, not sure if that's relevant here.

File: src/components/Forms/MaskedTextField.js (hunk 1 above)
Comment (loganmurdock) [reply src/components/Forms/MaskedTextField.js]: Actually just learned about this on Monday at workshop. Yeah, aria-labelledby and aria-describedby accept multiple ids just like class names. Pretty cool.

Comment (ychoi1) [reply src/components/Forms/MaskedTextField.js]: Correct, it'll read off in order of each element listed

File: src/components/Forms/TextFieldHint.js (hunk 2 above)
Comment (ychoi1) [reply src/components/Forms/TextFieldHint.js]: Right. When you tab around in the screen reader and the hint isn't showing, the hint will still receive focus by the screen reader and read off

File: src/components/Forms/MaskedTextField.js (hunk 1 above)
Comment (wfro) [inline src/components/Forms/MaskedTextField.js]: Which element gets the `error_<input id>` class? I can't find one at a glance but I remember it possibly being part of the form system?

File: src/components/Forms/docs/DateField.md [hunk 4]
```diff
@@ -0,0 +1,14 @@
+DateField Example:
```
Comment (wfro) [inline src/components/Forms/docs/DateField.md]: Is this related to the other accessibility stuff?

File: src/components/Forms/TextFieldHint.js [hunk 5]
```diff
@@ -35,17 +37,22 @@ class TextFieldHint extends Component {
...
   }
 
   render() {
-    const { disabled, show, style, text } = this.props
+    const { disabled, show, style, text, inputId } = this.props
 
     return (
-      <div style={[styles.root, style, disabled && styles.disabled, show && styles.show]}>
+      <div
+        id={`hint_${inputId}`}
```
Comment (wfro) [inline src/components/Forms/TextFieldHint.js]: more of a nit - since you have to define the hint_<x> class in the parent component anyways, you can pass in the full `hint_${inputId}` string instead of just `inputId`.

File: src/components/Forms/MaskedTextField.js (hunk 1 above)
Comment (ychoi1) [reply src/components/Forms/MaskedTextField.js]: `ValidationError`

File: src/components/Forms/TextFieldHint.js (hunk 2 above)
Comment (mmarcuccio) [reply src/components/Forms/TextFieldHint.js]: Also, this will interrupt the CSS transition since it happens immediately. You want to use multiple transitions, one on visibility and one on opacity.  The visibility transition should have a delay value equal to the duration of the opacity transition.

File: src/components/Forms/docs/DateField.md (hunk 4 above)
Comment (ychoi1) [reply src/components/Forms/docs/DateField.md]: I just added this since I think the Rx team added it, and didn't see an example listed for it.

File: src/components/Forms/MaskedTextField.js (hunk 3 above)
Comment (ychoi1) [reply src/components/Forms/MaskedTextField.js]: Good call, didn't see TextField using hint. Will update if needed

File: src/components/Forms/TextFieldHint.js (hunk 5 above)
Comment (mmarcuccio) [inline src/components/Forms/TextFieldHint.js]: It might be better for future reusability to apply the id prop directly here, and pass in `hint_${inputId}` from the parent component.

File: src/components/Forms/TextFieldHint.js (hunk 2 above)
Comment (ychoi1) [reply src/components/Forms/TextFieldHint.js]: Wouldn't that be ok since the opacity will still be transitioning while visibility is already set? From testing, it looks like opacity transition still takes effect regardless of visibility.

Comment (mmarcuccio) [reply src/components/Forms/TextFieldHint.js]: Before it would both fade in and fade out. After this change it will only fade in. Setting visible false happens immediately.

File: src/components/Forms/TextField.js [hunk 6]
```diff
@@ -252,7 +259,7 @@ class TextField extends React.Component {
             type={type}
             aria-required={required}
             aria-invalid={hasError}
-            aria-describedby={hasError ? `error_${inputId}` : null}
+            aria-describedby={[hasError ? `error_${inputId}` : null, hintText ? `hint_${inputId}` : null].join(' ').trim()}
```
Comment (ychoi1) [inline src/components/Forms/TextField.js]: Been having trouble getting this sort of logic to either be cleaner or follow complexity rules. Open to suggestions on refactoring this bit

Comment (mmarcuccio) [reply src/components/Forms/TextField.js]: Not that much better, but here's another approach:
```
aria-describedby={[hasError && `error_${inputId}`, hintText && `hint_${inputId}`].filter(Boolean).join(' ')
```

**Comment by ychoi1** [comment]
Updated with review changes.

Note: `Select` also needs accessibility changes. Will create separate branch for it since the changes aren't minimal.

**Comment by ychoi1** [comment]
Before merging, should we consider addressing the codeclimate issue?
